### PR TITLE
[runtime_env] Gracefully fail tasks when an environment fails to be set up

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

Partial solution to https://github.com/ray-project/ray/issues/16032.

This will fail all tasks with a given runtime_env if the environment fails to be set up.

Future extensions / follow-ups:
- If a named conda environment doesn't exist the setup hook does not currently fail.
- If the correct version of Ray isn't installed in the env the setup hook does not fail.
- We need an RPC to the agent every time a worker is started even if the env has already been setup or has failed.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by edoakes** [comment]
There are likely some unit tests that are going to fail due to changed interfaces. Will clean those up.

@ericl LMK if there are some C++ unit tests I should be updating as well. I'm mostly just reusing existing interfaces here.

File: dashboard/modules/runtime_env/runtime_env_agent.py
Comment (ericl) [inline dashboard/modules/runtime_env/runtime_env_agent.py:42]:
```suggestion
        if self._failed_count[serialized_env] > MAX_ATTEMPTS:
```

Retry a few times?

File: src/ray/protobuf/common.proto
Comment (ericl) [inline src/ray/protobuf/common.proto:118]:
```suggestion
  // Indicates that the GCS service failed to create the actor.
```

File: dashboard/modules/runtime_env/runtime_env_agent.py
Comment (ericl) [inline dashboard/modules/runtime_env/runtime_env_agent.py:73]: Oh do we already retry above?

Comment (edoakes) [reply dashboard/modules/runtime_env/runtime_env_agent.py:73]: Yep that's right, we retry creating the env 3 times by default before ever marking it failed. My change here is just caching the failure once we've already failed 3x so we don't go through the whole loop again. Effectively once an env fails 3x in a row, we assume it will never succeed in the future. We may want something more sophisticated in the future if we run into transient failures but this seems reasonable & simple to me for now.

File: src/ray/core_worker/transport/direct_task_transport.cc
Comment (architkulkarni) [inline src/ray/core_worker/transport/direct_task_transport.cc:512]: Just making sure I understand: the reason it's okay to fail all the tasks in the queue is that the runtime env hash is part of the scheduling key, so every task in the queue is trying to use the same runtime env, right?

**Review by architkulkarni — APPROVED** [review]
Thanks for the PR, looks good to me! Currently the runtime env setup for the Ray Client server doesn't go through the agent, so this PR won't affect it, but for that case we don't have the issue of tasks repeatedly trying to start so it should be okay.

File: src/ray/core_worker/transport/direct_task_transport.cc
Comment (edoakes) [reply src/ray/core_worker/transport/direct_task_transport.cc:512]: Yep, that's right.

File: python/ray/exceptions.py
Comment (simon-mo) [inline python/ray/exceptions.py:277]: where does this happen?

Comment (edoakes) [reply python/ray/exceptions.py:277]: They get streamed through the raylet (I'm not actually sure how this works)

Comment (edoakes) [reply python/ray/exceptions.py:277]: This existed before my change

Comment (simon-mo) [reply python/ray/exceptions.py:277]: nvm it's in agent_manager.cc

**Comment by edoakes** [comment]
@ericl could you approve this please? Need code owner approval for proto change.

**Review by ericl — APPROVED** [review]

**Pull request merged**
